chain-impl-mockchain/src/setting.rs

```diff
@@ -159,6 +303,160 @@ impl Readable for UpdateProposal {
     }
 }
 
+#[derive(Clone, Debug)]
+pub struct UpdateProposalWithProposer {
+    pub proposal: UpdateProposal,
+    pub proposer_id: UpdateVoterId,
+}
+
+impl HasPublicKeys for UpdateProposalWithProposer {
+    fn public_keys<'a>(
+        &'a self,
+    ) -> Box<ExactSizeIterator<Item = &PublicKey<Ed25519Extended>> + 'a> {
+        Box::new(std::iter::once(&self.proposer_id.0))
+    }
+}
+
+impl property::Serialize for UpdateProposalWithProposer {
+    type Error = std::io::Error;
+    fn serialize<W: std::io::Write>(&self, writer: W) -> Result<(), Self::Error> {
+        use chain_core::packer::*;
+        let mut codec = Codec::from(writer);
+        self.proposal.serialize(&mut codec)?;
```

Comment:
minor point, but I don't think you don't need to create a codec and can pass the writer directly here